### ISSUE TITLE
Suggestion: Expect should also handle promises

### ISSUE THREAD

**Issue by danielalves** [jasmine/jasmine#1447]
## Expected Behavior
Currently Jasmine supports `async` code in `before*`, `after*` and `it` blocks. I think it would be a great addition to the framework if the `expect` function could handle promises as well. The main reasons that come to my mind are:
  - Code would be less verbose
  - We would be able to use some existent matchers, improving the console output

## Current Behavior
Currently, if one wants to test if an `async` function throws, this is one possible way out:
```js
it('throws an error on invalid argument', async () => {
  let error;
  try {
    await methodBeingTested(invalidArgument);
  } catch (err) {
    error = err;
  }
  expect(error).toEqual(ExpectedError);
});
```

If `expect` did handle `async` code, one could just do:
```js
it('throws an error on invalid argument', async () => {
  expect(async () => {
    await methodBeingTested(invalidArgument);
  }).toThrow(ExpectedError);
});
```

## Possible Solution
I don't know how Jasmine's internals work, so I think any solution I could provide here would be too naive. =/

## Context
There's much `async` code to be tested nowadays. Async `before*`, `after*` and `it` blocks handle almost every situation I can think of, except by the case described above. If `expect` could handle `async` code as well, I think the framework would be more feature complete. =)

**Comment by sgravrock** [comment]
We've had a few requests for this feature. The biggest problem is that async is "contagious" -- if a function calls an async function and wants to do anything with the result, it has to be async itself. So if matchers are async, then all specs have to be async as well. Which means that implementing this feature would break the majority of Jasmine specs.

The other problem is that the feature would be difficult to use correctly. In order for Jasmine to associate the failure with the spec that causes it, it would be necessary to `await` the `expect` call so that the spec doesn't finish until after the matcher is evaluated:

```
it('throws an error on invalid argument', async () => {
  await expect(async () => {
    await methodBeingTested(invalidArgument);
  }).toThrow(ExpectedError);
})
```

That would be really easy to forget, in which case Jasmine would associate the failure with the wrong spec or no spec at all. We already see that problem when people call `done` at the wrong time, and it can be incredibly time-consuming to debug.

So, to sum up: I recognize the value of this feature, but it would require users to rewrite most existing specs and would make writing Jasmine specs much more error prone. I'm open to suggestions if anyone has a clever idea for how to make this work without requiring specs to be `async` or to `await` the call to `expect`. But I'm going to go ahead and close this as it's not something that we are likely to implement.

**Comment by danielalves** [comment]
Hey, @sgravrock! Thanks for your answer!

After reading all your points above, I came up with an idea. Please, tell me if it could work (and again, forgive me if this is too naive =) ).

First of all, we would need to borrow the idea of cancelable promises described on [React's Blog](https://facebook.github.io/react/blog/2015/12/16/ismounted-antipattern.html) and alter it a bit:


```js
function makeCancelablePromise(originalPromise) {
  let resolved = false;
  let hasCanceled = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    originalPromise.then(val => {
      resolved = true;
      (hasCanceled ? reject({ isCanceled: true }) : resolve(val))
    });
    originalPromise.catch(error => {
      resolved = true;
      (hasCanceled ? reject({ isCanceled: true }) : reject(error))
    });
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled = true;
    },
    isResolved() {
      return resolved;
    },
  };
}
```

Ok. Now, the goal is to use `expect` with `async` functions without needing to use `await` before it:

```js
// `It`'s callback wouldn't need to be async here
it('throws an error on invalid argument', () => {
  expect(async () => {
    await methodBeingTested(invalidArgument);
  }).toThrow(ExpectedError);
});
```

The rough idea I came up with goes like this:

```js
function expect(param, ...) {

  // ...

  if (isPromise(param)) {
      const cancelablePromise = makeCancelablePromise(promise);

      // We will need this to make the parent it wait for all
      // async expects before exiting
      parentIt.registerAsyncExpect(cancelablePromise.promise);

      // This will let us abort async expects after the
      // specified timeout
      const timeoutId = setTimeout(() => {
          if (!cancelablePromise.isResolved()) {
            cancelablePromise.cancel();
          
            // Fails the expect here telling the user that the 
            // promise was not called before the specified timeout
            fail(...);
          }
      }, jasmine.DEFAULT_TIMEOUT_INTERVAL);
  }

  // ...
}

function it(...) {

  // ...

  callTestCallback();

  // Either an async expect will get called or it'll get canceled, so
  // there's no need to track the async expects timeouts
  await Promise.all(this.asyncExpects);

  // ...
}

```

Where:

- `parentIt`: the `it` block where `expects` get called
- `registerAsyncExpect`: a method that would keep track of every `async` callback passed to expects. In the code above, it is filling the `this.asyncExpects` array.

One problem to be worked out is how to abort on a first failure. Maybe listening for a first rejection could work. We would also need to keep what matcher should be called and use it inside `makeCancelablePromise`'s `wrappedPromise`'s handlers.

**Comment by sgravrock** [comment]
It's not as "easy" as that, unfortunately. I can see a few challenges that might not be immediately obvious:

* Jasmine can't create promises because it needs to run in environments that don't have them. (And we can't use a polyfill because that could cause users' tests to falsely pass in environments that lack promise support.) So you'd need a purely callback-based equivalent of `makeCancelablePromise`.
* Jasmine would need to support arbitrary numbers of these async `expect` calls in a given spec, which would add more complexity.
* `expect` can't just decide to be async based on the type of the parameter because that would break the behavior of invocations like `expect(promiseThatNeverResolves).toEqual(jasmine.any(Promise))`. Or imagine a custom matcher like this: `expect(somePromise).not.toBeSettled()`. Any implementation would need to preserve those behaviors.
* It seems like this might significantly complicate the code that determines when a spec is done, which is already pretty challenging to work with.

You are of course welcome to take a crack at this. But one of the key criteria we use when evaluating PRs for new features is whether the feature is worth the impact on maintainability. In this case that may be a high bar to clear.

**Comment by darlanalves** [comment]
@sgravrock 
What would be the chances of introducing a new async expect function that has all the special cases covered and yet doesn't interfere with current specs in the wild?
I don't have a clear answer to that either, but it's painful for a lot of developers including me...

How about an extension of expect() that handles promises?
e.g. `expect.async(promise).toBeResolved()`

**Comment by brunolm** [comment]
If anyone got here from Google, you can use this:

```ts
const myAsyncFunction = async () => { ... }
expectAsync(myAsyncFunction()).toBeRejected()
```

**Comment by slackersoft** [comment]
You should also be able inline the async function, something like:

```javascript
expectAsync(async () => { ... }).toBeRejected();
```
